File: extScript/mazeGenerator.py
import pdb
import random

class MazeNode:
	def __init__(self, InVisited = False):
		self.wall = []
		self.isVisited = InVisited
		self.nextWay = [(-1, 0, 'N'), (1, 0, 'S'), (0, -1, 'W'), (0, 1, 'E')]

	def addWall(self, InWay):
		self.wall.append(InWay)

	def prevRoute(self, InDir):
		if InDir == 'N':
			del self.nextWay[1]
		if InDir == 'S':
			del self.nextWay[0]
		if InDir == 'W':
			del self.nextWay[3]
		if InDir == 'E':
			del self.nextWay[2]

	def __str__(self):
		r = ''
		if self.isVisited:
			r = 'T'
		else:
			r = 'F'
		return r + ''.join(self.wall)



class Maze:
	def __init__(self, InSize):
		self.mapSize = InSize + 2
		self.mapNode = {}

		for c in range(self.mapSize):
			for r in range(self.mapSize):
				if r == 0 or r == self.mapSize-1 or c == 0 or c == self.mapSize-1:
					self.mapNode[(c, r)] = MazeNode(True)
				else:
					self.mapNode[(c, r)] = MazeNode()

	def show(self):
		print(self.mapSize)
		for i in range(self.mapSize):
			print(f'{i:<7}',end='')
		print()
		for c in range(self.mapSize):
			for r in range(self.mapSize):
				print(f'{self.mapNode[(c, r)]}'.ljust(6), end=' ')
			print()

	def generate(self):
		dfs = [(1, 1)]
		while dfs:
			if not self.mapNode[dfs[-1]].nextWay:
				dfs.pop()
				continue
			#print(dfs)
			#print(dfs[-1])
			#print(self.mapNode[dfs[-1]].nextWay)
			#m.show()
			#pdb.set_trace()
			self.mapNode[dfs[-1]].isVisited = True
			random.shuffle(self.mapNode[dfs[-1]].nextWay)

			while self.mapNode[dfs[-1]].nextWay:
				nw = self.mapNode[dfs[-1]].nextWay.pop()
				npos = (dfs[-1][0] + nw[0], dfs[-1][1] + nw[1])
				if self.mapNode[npos].isVisited:
					self.mapNode[dfs[-1]].addWall(nw[2])
				else:
					dfs.append(npos)
					self.mapNode[dfs[-1]].prevRoute(nw[2])
					break

m = Maze(4)
m.generate()
m.show()
pdb.set_trace()

File: Source/ProjectM/MazeGenerator.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "MazeGenerator.h"
#include <random>
#include <Engine/World.h>
#include "Global.h"
#include "WorldBuilder.h"


class MazeGenerator::MazeImpl {
public:
    MazeImpl() = default;

    ~MazeImpl() = default;

    void init(uint8 InMapSize)
    {
        PrintLine();

        _mapSize = InMapSize + 2;

        _dfs.push_back({ 1, 1 });

        _startPos = Position(0, 0);
        _endPos = Position(0, 0);
    }

    void generate()
    {
        PrintLine();

        std::random_device rd;
        auto rng = std::default_random_engine{ rd() };

        _mapNode.clear();

        for (uint8 c = 0; c < _mapSize; c++)
        {
            for (uint8 r = 0; r < _mapSize; r++)
            {
                Position pos{ c, r };
                MazeNode* mn = new MazeNode();
                if (r == 0 || r == _mapSize - 1 || c == 0 || c == _mapSize - 1)
                {
                    mn->setVisited();
                }
                _mapNode[pos] = mn;
            }
        }

        uint16 goalDepth = 0;

        while (!_dfs.empty())
        {
            if (goalDepth < _dfs.size())
            {
                goalDepth = _dfs.size();
                _endPos = _dfs.back();
            }


            if (_mapNode[_dfs.back()]->nextWay.empty())
            {
                _dfs.pop_back();
                continue;
            }

            _mapNode[_dfs.back()]->setVisited();
            //int random = rand() % _mapNode[_dfs.back()]->nextWay.size();
            //Way sel_elem = _mapNode[_dfs.back()]->nextWay[random];
            //_mapNode[_dfs.back()]->nextWay.erase(_mapNode[_dfs.back()]->nextWay.begin() + random);
            std::shuffle(_mapNode[_dfs.back()]->nextWay.begin(), _mapNode[_dfs.back()]->nextWay.end(), rng);

            while (!_mapNode[_dfs.back()]->nextWay.empty())
            {
                Way nw = _mapNode[_dfs.back()]->nextWay.back();
                _mapNode[_dfs.back()]->nextWay.pop_back();

                Position npos{ uint8(_dfs.back().Col + nw.Pos.Col), uint8(_dfs.back().Row + nw.Pos.Row) };
                if (_mapNode[npos]->isVisited())
                {
                    _mapNode[_dfs.back()]->addWall(nw.Dir);
                }
                else
                {
                    _dfs.push_back(npos);
                    _mapNode[_dfs.back()]->prevRoute(nw.Dir);
                    break;
                }
            }
        }
    }

    void build()
    {
//         AWorldBuilder* worldBuilder = _world->SpawnActor<AWorldBuilder>(AWorldBuilder::StaticClass());
//         worldBuilder->build(_mapNode);
    }

    std::map<Position, MazeNode*> getMaze()
    {
        return _mapNode;
    }

    Position getGoalPos()
    {
        return _endPos;
    }

private:
    uint8 _mapSize;
    std::vector<Position> _dfs;
    std::map<Position, MazeNode*> _mapNode;
    Position _startPos;
    Position _endPos;
};

MazeGenerator::MazeGenerator() : pimpl_(new MazeImpl) {};
MazeGenerator::~MazeGenerator() = default;

void MazeGenerator::init(uint8 InMapSize)
{
    pimpl_->init(InMapSize);
}

void MazeGenerator::generate()
{
    pimpl_->generate();
}

void MazeGenerator::build()
{
    pimpl_->build();
}

Position MazeGenerator::getGoalPos()
{
    return pimpl_->getGoalPos();
}

std::map<Position, MazeNode*> MazeGenerator::getMaze()
{
    return pimpl_->getMaze();
}

